backend/src/packages/chaiNNer_external/external_stable_diffusion/automatic1111/upscaling.py

```diff
@@ -0,0 +1,201 @@
+            ),
+        ),
+    ],
+    outputs=[
+        ImageOutput(
+            image_type="""
+                def nearest_valid(n: number) = int & floor(n);
+
+                let in_w = Input0.width;
+                let in_h = Input0.height;
+                let ratio_w = width/in_w;
+                let ratio_h = height/in_h;
+                let larger_ratio = if ratio_w>ratio_h { ratio_w } else { ratio_h };
+
+                let mode = Input1;
+                let factor = Input2;
+
+                let crop = Input5;
+                let width = Input3;
+                let height = Input4;
+
+                match mode {
+                    UpscalerMode::ScaleTo => if crop {
+                        Image { width: width, height: height }
+                    } else {
+                        Image {
+                            width: nearest_valid(in_w*larger_ratio),
+                            height: nearest_valid(in_h*larger_ratio)
+                        }
+                    },
+                    UpscalerMode::ScaleBy => Image {
+                        width: nearest_valid(in_w*factor),
+                        height: nearest_valid(in_h*factor)
+                    }
+                }
+            """,
+            channels=3,
+        )
+    ],
+)
+@cached
+def upscale_node(
+    image: np.ndarray,
+    mode: UpscalerMode,
+    upscaling_resize: float,
+    width: int,
+    height: int,
+    crop: bool,
+    upscaler_1: Enum,
+    use_second_upscaler: bool,
+    upscaler_2: Enum,
+    upscaler_2_visibility: float,
+) -> np.ndarray:
+    if mode == UpscalerMode.SCALE_BY:
+        resize_mode = 0
+    else:
+        resize_mode = 1
+
+    if use_second_upscaler:
+        u2 = upscaler_2.value
+    else:
+        u2 = "None"
+
+    request_data = {
+        "resize_mode": resize_mode,
+        "show_extras_results": False,
+        "gfpgan_visibility": 0.0,
+        "codeformer_visibility": 0.0,
+        "codeformer_weight": 0.0,
+        "upscaling_resize": upscaling_resize,
+        "upscaling_resize_w": width,
+        "upscaling_resize_h": height,
+        "upscaling_crop": crop,
+        "upscaler_1": upscaler_1.value,
+        "upscaler_2": u2,
+        "extras_upscaler_2_visibility": upscaler_2_visibility,
+        "upscale_first": False,
+        "image": encode_base64_image(image),
+    }
+    response = post(
+        path=STABLE_DIFFUSION_EXTRA_SINGLE_IMAGE_PATH, json_data=request_data
+    )
+    result = decode_base64_image(response["image"])
+
+    ih, iw, _ = get_h_w_c(image)
+    rh, rw, _ = get_h_w_c(result)
+    ratio_w = width / iw
+    ratio_h = height / ih
+    if ratio_w > ratio_h:
+        larger_ratio = ratio_w
+    else:
+        larger_ratio = ratio_h
+
+    if mode == UpscalerMode.SCALE_TO:
+        if crop:
+            assert (rw, rh) == (
+                width,
+                height,
+            ), f"Expected the returned image to be {width}x{height}px but found {w}x{h}px instead "
```

Comment:
I agree with the linter. Those should be `rw` and `rh`. Same for the other 2 error messages.

```suggestion
            ), f"Expected the returned image to be {width}x{height}px but found {rw}x{rh}px instead "
```

Reply:
For sure, copy-paste error.